File: Data Structures & Algorithms/non-overlapping-intervals/submission-17.cs
public class Solution {
    private int[] memo;

    public int EraseOverlapIntervals(int[][] intervals) {
        Array.Sort(intervals, (a, b) => a[1].CompareTo(b[1]));
        int n = intervals.Length;
        memo = new int[n];  
        Array.Fill(memo, -1);

        int maxNonOverlapping = Dfs(intervals, 0);
        return n - maxNonOverlapping;
    }

    private int Dfs(int[][] intervals, int i) {
        if (i >= intervals.Length) return 0;
        if (memo[i] != -1) return memo[i];

        int res = 1;
        for (int j = i + 1; j < intervals.Length; j++) {
            if (intervals[i][1] <= intervals[j][0]) {
                res = Math.Max(res, 1 + Dfs(intervals, j));
            }
        }
        memo[i] = res;
        return res;
    }
}

File: Data Structures & Algorithms/maximum-subarray/submission-17.java
class Solution {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n + 1][2];
        
        for (int i = n - 1; i >= 0; i--) {
            dp[i][1] = Math.max(0, nums[i] + dp[i + 1][1]);
            dp[i][0] = Math.max(dp[i + 1][0], nums[i] + dp[i + 1][1]);
        }
        
        return dp[0][0];
    }
}